### ISSUE TITLE
ERROR: Response status code does not indicate success: 404 (Not Found).

### ISSUE THREAD

**Issue by Antobite** [lay295/TwitchDownloader#396]
### Checklist

- [X] I have checked the [issue page](https://github.com/lay295/TwitchDownloader/issues) for duplicates
- [X] I am running the latest version ([download here](https://github.com/lay295/TwitchDownloader/releases/latest))

### Describe your issue here

I am running v1.50.5, whenever I try and chat render a json I just downloaded through the program, I get this error message. I have restarted, reinstalled, cleared cache in the config menu, set default settings, I am still getting this error.

![image](https://user-images.githubusercontent.com/70809729/199028282-e1ed8768-86c2-43e9-9ffa-ee414eafecc6.png)

**Comment by ScrubN** [comment]
I have no clue what could have returned 404. Can you provide the VOD/Clip ID or provide the json if the video expired.

**Comment by Antobite** [comment]
> https://www.twitch.tv/videos/209227327

**Comment by Antobite** [comment]
Actually I'm wondering if there's some issue with downloading the json in the first place. I've tried 3 different times to download the chat and the file sizes are slightly different. I also can't open any of the three in Notepad of Notepad++, both programs just stop responding.

This is the same vod I linked above, 3 attempts to download.

![image](https://user-images.githubusercontent.com/70809729/199184167-34bbf194-e426-47f0-a276-e5f48b7a664f.png)

**Comment by ScrubN** [comment]
The issue is regarding emote downloading. Does the streamer have the "modify an emote" redeem? I suspect that might be it but I'm still digging

**Comment by Antobite** [comment]
> The issue is regarding emote downloading. Does the streamer have the "modify an emote" redeem? I suspect that might be it but I'm still digging

I have no idea, the vod is from 2017 if that was even a thing back then.

**Comment by ScrubN** [comment]
Honestly I should've checked the upload date. That probably explains everything, though reading the code I figured it would just skip over every missing emote. Again, still digging but deleted emotes is probably the case.

**Comment by ScrubN** [comment]
The emote being removed from twitch is indeed the cause of the 404. There's supposed to be a a list of failed emotes to ignore but for whatever reason the list is never populated after initialization. I'm fixing that right now and then the issue will be gone when it's merged

**Comment by Stargateur** [comment]
> I also can't open any of the three in Notepad of Notepad++, both programs just stop responding.

It's not surprise you can't open a 180 MiO text file in notepad. You may try using vscode or atom but don't expect it's will "work". Thus you can use tool like `more` in console that will show you the file for sure.

**Comment by ScrubN** [comment]
> > it would just skip over every missing emote.
> 
> Missing emote should maybe have a "missing" emote like a black square or something ? Maybe configurable color ?

Good idea. I've posed the idea in the comments of the fix PR (#399)

**Comment by Antobite** [comment]
> The emote being removed from twitch is indeed the cause of the 404. There's supposed to be a a list of failed emotes to ignore but for whatever reason the list is never populated after initialization. I'm fixing that right now and then the issue will be gone when it's merged

Awesome, thank you. If the info helps at all, I tried 1.40.7 and it's started to render the video without issues. Also instead of a black square, just the original keyword for the emote would be the best solution (for example, just the word 'PogChamp' instead of the PogChamp emote itself).

**Comment by glubsy** [comment]
> for example, just the word 'PogChamp' instead of the PogChamp emote itself).

Why not :PogChamp: without the actual emote?

**Comment by ScrubN** [comment]
> Awesome, thank you. If the info helps at all, I tried 1.40.7 and it's started to render the video without issues. Also instead of a black square, just the original keyword for the emote would be the best solution in my opinion (for example, just the word 'PogChamp' instead of the PogChamp emote itself).

Reading into the rendering code I found this section. Writing the emote keyword in the place of failed emotes was already in place, there was just the bug in the emote fetching.

```cs
//First party emote
string emoteId = fragment.emoticon.emoticon_id;
if (emoteList.Any(x => x.Id == emoteId))
{
    TwitchEmote twitchEmote = emoteList.First(x => x.Id == emoteId);
    if (drawPos.X + twitchEmote.Width > renderOptions.ChatWidth - renderOptions.SidePadding)
        AddImageSection(sectionImages, ref drawPos, ref defaultPos);
    Point emotePoint = new Point();
    emotePoint.X = drawPos.X;
    emotePoint.Y = (int)(sectionImages.Sum(x => x.Height) - renderOptions.SectionHeight + ((renderOptions.SectionHeight - twitchEmote.Height) / 2.0));
    emotePositionList.Add((emotePoint, twitchEmote));
    drawPos.X += twitchEmote.Width + renderOptions.EmoteSpacing;
}
else
{
    //Probably an old emote that was removed
    DrawText(fragment.text, messageFont, true, sectionImages, ref drawPos, ref defaultPos);
}
```

**Comment by ScrubN** [comment]
Incase you were wondering how it looks, I found a message containing deleted emotes. On the left is the json showing the emotes and their ids, on the right is the rendered output. I'm happy with that as it is.

![1](https://user-images.githubusercontent.com/72096833/199412906-eeec55e6-4b5f-47f1-b1f2-3e3a612f1816.jpg)

> Why not :PogChamp: without the actual emote?
`:emote:` isn't how emotes work on twitch; that's how you inline search for emotes. In my opinion it would be weird to show the inline search/discord emote syntax rather than the raw text version

**Comment by Zibbp** [comment]
> would it be possible for the user to somehow locally add an emote to the json file, so someone could theoretically fix deleted emotes themselves?

The embed emotes can be added / modified at the bottom of the chat json. The structure can be seen below. It may be beneficial to download any chat with emotes embedded so you can copy the structure. The Base64 string of the image can be found a variety of ways including this website https://codebeautify.org/image-to-base64-converter. Gifs also working in this format.

```json
{
"streamer": {},
"comments": [],
"video": {},
"emotes": {
  "thirdParty": [
     {
        "id": "54fa8f1401e468494b85b537",
        "imageScale": 2,
        "data": "Base64 String of Image",
        "name": ":tf:",
        "url": null,
        "width": 28,
        "height": 28
      }
   ],
   "firstParty": [
     {
        "id": "555555584",
        "imageScale": 2,
        "data": "Base64 String of Image",
        "name": null,
        "url": null,
        "width": 24,
        "height": 18
      },
    ],
}
}
```

**Comment by cr08** [comment]
I also ran into the same thing just now and seems like potentially identical issues. Getting the 404 error on an old highlight. If I turn off emote embeds the json downloads fine.

https://www.twitch.tv/videos/834170247

Running CLI version 1.50.5 right now.

**Comment by ScrubN** [comment]
> I also ran into the same thing just now and seems like potentially identical issues. Getting the 404 error on an old highlight. If I turn off emote embeds the json downloads fine.
> 
> https://www.twitch.tv/videos/834170247
> 
> Running CLI version 1.50.5 right now.
> 
> From the POV of trying to use the CLI in a script, would it be possible to just have it ignore the embed emotes flag in these scenarios and continue with the download rather than failing?

I wish you read the thread. There is already a fix waiting to be merged: #399
